Reduce TLB misses during radix partitioning

Greenscreen23 opened the pull request.

Bouncner commented:
Can you run radix bits starting from 1?
And since we partition to keep a hash table in L2 (usually 500KB'ish or  larger) radix bits that lead to partitions < 10,000 elements are rather distracting. We don't have to care for that.

Right now I think:
 * for very large joins, the changes looks very good
 * for "typical" SF 10 joins (mostly the left two plots), we need more data on the lower bits

Greenscreen23 commented:
Few radix bits and with a specified size per radix partition and not a specified total size. For <= 5 radix bits, Master is faster than the current version. There may be some substantial noise in this data still.
![micro2](https://github.com/user-attachments/assets/972f140c-3de5-44e3-b8e1-a100f52df54e)

Bouncner commented:
Can you add error bars to be sure we don't discuss noise?

Seeing the data I am thinking of templating the writing with a "fast path" for bits<=6. What do you think?

Bouncner commented on src/test/lib/operators/join_hash/join_hash_steps_test.cpp:
```diff
@@ -328,4 +342,68 @@ TEST_F(JoinHashStepsTest, ThrowWhenNoNullValuesArePassed) {
+}
+
+TEST_F(JoinHashStepsTest, PartitionLargerDatasetWithNullsCorrectly) {
+  const size_t radix_bit_count = 3;
```
auto auto auto

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -516,9 +520,75 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+  void prefetch(size_t partition_id) {
+    const auto* const ptr = reinterpret_cast<char*>(data.data() + partition_id);
+    for (auto cacheline_idx = size_t{0}; cacheline_idx < bucket::CACHELINES_PER_STORE; ++cacheline_idx) {
+      __builtin_prefetch(ptr + (cacheline_idx * BYTES_PER_CACHELINE), 1 /* = write */, 3 /* = highest priority */);
```
It's locality, not priority.

Bouncner replied:
I am curious for your decision to use 3. First I though, it makes sense as we are probably going to write to this array again soon.
But I checked the Schuhknecht paper (btw, is it linked in the comments somewhere?) and they use a value of 0.

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -516,9 +520,75 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+  }
+
+  void prefetch(size_t partition_id) {
+    const auto* const ptr = reinterpret_cast<char*>(data.data() + partition_id);
```
`partition_pointer` or something like that?

Bouncner replied:
std::byte instead of char?

Greenscreen23 commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -516,9 +520,75 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+// elements. You should consider enabling it if the number of radix buckets is larger than the size of your TLB. A good
+// rule of thumb seems to be starting at about 7 radix bits. This setting will be ignored if T is not trivially
+// copy assignable, as we have to copy the elements bitwise when doing chunked copy.
+template <typename T, typename HashedType, bool keep_null_values, bool reduce_tlb_pressure = false>
```
But both prefetching and nt stores arise from the solution originally targeted at reducing tlb pressure. We don't prefetch the input tuples every and we don't really use nontemporal writes when doing the normal "non tlb reducing" algorithm. So I think it would be missleading to mention prefetching and nt

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -516,9 +520,83 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+  requires std::is_trivially_copy_assignable_v<T>
+union TemporaryRadixBucket {
+  static constexpr auto BYTES_PER_ELEMENT = sizeof(PartitionedElement<T>);
+  // We want the size of this container to a multiple of the element size (so we don't copy uninitialized bytes) and
```
```suggestion
  // We want the size of this container to be a multiple of the element size (so we don't copy uninitialized bytes) and
```

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -516,9 +520,83 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+
+  struct {
+    std::array<PartitionedElement<T>, ELEMENTS_PER_STORE - 1> elements;
+    // Every column type T is at least 4 bytes and its RowID is at least 8, so we have space for these two numbers
```
```suggestion
    // Every column type T is at least 4 bytes and its RowID is at least 8, so we have space for these two numbers.
```

Greenscreen23 commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -684,17 +678,58 @@ RadixContainer<T> partition_by_radix(const RadixContainer<T>& radix_container,
+              __builtin_prefetch(&output[radix].elements[output_idx + 1], 1, locality);
+            } else if constexpr (variant == 'D') {
+              __builtin_prefetch(
+                  reinterpret_cast<std::byte*>(&output[radix].elements[output_idx]) + BYTES_PER_CACHELINE, 1, locality);
```
Yes, on others as well. That is why I switched to `elements.data() + index` instead of `&elements[index]`. This should be fine for the evaluation, but we should think about this before merging.

We are just prefetching, so we don't read any out-of-bounds data, but I am not sure if there is an issue if we prefetch a cacheline that's on a virtual page that is not mapped to a physical page.

Greenscreen23 commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -52,8 +52,8 @@ The original value is used to detect hash collisions.
 struct PartitionedElement {
-  RowID row_id;
   T value;
+  RowID row_id;
```
If an element is between two cachelines, then we can be sure that `value` is still on the old cacheline.

This only really concerns the input, which we read sequentially, so we probably already have the next cacheline, so I could also revert it, if you want.

Greenscreen23 commented:
Microbenchmark results on AMD:
Variants:
A: Prefetch the next items cacheline if it is the first item fully on a new cacheline
B: Prefetch the next items cacheline if it is at least partially on a new cacheline (if it is on two cachelines, prefetch the second one)
C: Always prefetch the start of the next item, no matter its cacheline alignment
D: Always prefetch the next cacheline
N: Don't do anything
![micro-prefetch-amd](https://github.com/user-attachments/assets/c56b0697-0065-47b8-ada5-b61f178a76b7)
Aggregated for the highest possible size of all radix bit counts:
```
Variant
'A'    11788.945391
'B'    12009.938582
'C'    11759.807855
'D'    12138.476518
'N'    11739.167155
dtype: float64
```

Bouncner commented:
> Microbenchmark results on AMD: Variants: A: Prefetch the next items cacheline if it is the first item fully on a new cacheline B: Prefetch the next items cacheline if it is at least partially on a new cacheline (if it is on two cachelines, prefetch the second one) C: Always prefetch the start of the next item, no matter its cacheline alignment D: Always prefetch the next cacheline N: Don't do anything ![micro-prefetch-amd](https://private-user-images.githubusercontent.com/43916057/539020287-c56b0697-0065-47b8-ada5-b61f178a76b7.svg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjkyMDEwMzksIm5iZiI6MTc2OTIwMDczOSwicGF0aCI6Ii80MzkxNjA1Ny81MzkwMjAyODctYzU2YjA2OTctMDA2NS00N2I4LWFkYTUtYjYxZjE3OGE3NmI3LnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTIzVDIwMzg1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0MTJiMTY0ZDNlNjVjNTQwNjE1MzBjMjI3MWJiZTg4ZTE4OGNmZWUxMTZlZWMzNWIxNmZiZTVhNmQ1Zjg3NDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.D2PJ21miy67XHT-bFDG1j2_SjN8zU-VnoX1PXhFph7k) Aggregated for the highest possible size of all radix bit counts:
> 
> ```
> Variant
> 'A'    11788.945391
> 'B'    12009.938582
> 'C'    11759.807855
> 'D'    12138.476518
> 'N'    11739.167155
> dtype: float64
> ```

What are the items on the x axis? The number of items per partition or overall?

Do you also prefetch for the non-TLB-optimized path? If so, can you also plot smaller radix bit counts?

Greenscreen23 commented:
> What are the items on the x axis? The number of items per partition or overall?

The number of items per partition.


> Do you also prefetch for the non-TLB-optimized path? If so, can you also plot smaller radix bit counts?

Because the number of radix bits is always high, this is always the micro buffer path. We can also test this for smaller radix bit counts, but that would be a different code path. For this code path, we should use tests with many radix bits to decide.

I am a bit unhappy about the amount of noise and the resulting ambiguity. Is not prefetching really the best option?

Greenscreen23 commented:
- Run on nemea
- 100 iterations
~- Check governor (and use max for measurements)~
The governor on hpc machines seems to be schedutil, on nemea its powersave

Greenscreen23 commented:
Microbenchmark results on Nemea (100 runs, performance governor):
Variants:
A: Prefetch the next items cacheline if it is the first item fully on a new cacheline
B: Prefetch the next items cacheline if it is at least partially on a new cacheline (if it is on two cachelines, prefetch the second one)
C: Always prefetch the start of the next item, no matter its cacheline alignment
D: Always prefetch the next cacheline
N: Don't do anything
![micro-prefetch-nemea](https://github.com/user-attachments/assets/dedb0fad-2f80-494e-9392-888242a79a68)
The error bars are stddev, not confidence intervals, which is why they don't shrink with the larger iteration count.

Aggregated (geo mean) for the highest item count:
```
Variant
'A'    3314.080840
'B'    3387.191295
'C'    3249.500291
'D'    3632.669353
'N'    3232.363511
```

Greenscreen23 commented:
I feel like the picture from nemea and 100 runs is a bit clearer. The two best options seem to be not to prefetch or to always prefetch the next item (even if it might be on the same cacheline). This might either imply that prefetching the next item is as good as not prefetching, or that prefetching a cacheline that is already in cache is cheap (on this specific cpu). There is a very light performance boost for prefetching the next item, which might hint towards the latter. I will do 100 runs on AMD as well to see if this pattern holds there as well.

Greenscreen23 commented:
When run on the HPC cluster (cx20)
![micro-prefetch-amd](https://github.com/user-attachments/assets/6c17ca41-85bf-4d44-936d-0f8d331f85a6)
```
Variant
'A'    2594.314769
'B'    2610.624306
'C'    2613.410127
'D'    2653.516275
'N'    2654.222779
```

Greenscreen23 commented:
Edit: I changed the aggregated values in the last two runs (prefetch on Nemea and AMD) to geomean.

The results from AMD and Nemea are quite different, which makes it difficult to select the best version. C seems to be have the most consistent, whereas A is a favorite for AMD and N is a favorite for Intel.  I aggregated the values across intel and amd (geomean again) in this table:

```
Variant
'A'    2932.19523035
'B'    2973.66506584
'C'    2914.1511574
'D'    3104.72981914
'N'    2929.06347847
```

Given that C wins here and given that C is also what Schuhknecht et al. did, I would go with C (always prefetch the next item)

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -520,9 +524,82 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
-template <typename T, typename HashedType, bool keep_null_values>
+// This is the cacheline size for most of the systems Hyrise usually runs on. If a system has larger cachelines (e.g.
+// Apple's M processors or IBM Power), using the default value will still run fine. Our implementation assumes that
+// this is a power of two.
```
> Our implementation assumes that this is a power of two.

Rather use a static_assert?

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -520,9 +524,82 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+// Apple's M processors or IBM Power), using the default value will still run fine. Our implementation assumes that
+// this is a power of two.
+constexpr auto BYTES_PER_CACHELINE = size_t{64};
+// This is a tunable parameter. Increasing it causes more cache usage during radix partioning, but less TLB pressure.
```
> [..]  it causes more cache usage [..]

Cache usage means memory usage? Or accesses to the cache?

Greenscreen23 replied:
I reformulated it. Is this better?

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -520,9 +524,82 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+
+// TemporaryRadixBucket is one "chunk" of data that can be copied to one specific partition once full. It is
+// implemented as a union so the number of items contained and the output position can be stored here while the bucket
+// is not full (called a micro row layout). This way they don't have to be fetched from some other location. The
```
`a micro row layout, see https://xxxxxx`. Can you please reference the paper here?

Bouncner replied:
> [..] can be stored here [..]

You  mean in the same cache line, right? We can always store output pod and count in the same struct, but it might be on another cache line then.

Greenscreen23 replied:
Is it better now?

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -520,9 +524,82 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+// TemporaryRadixBucket is one "chunk" of data that can be copied to one specific partition once full. It is
+// implemented as a union so the number of items contained and the output position can be stored here while the bucket
+// is not full (called a micro row layout). This way they don't have to be fetched from some other location. The
+// alter is to store the indices in a separate array (called micro column layout), which is desirable when this
```
```suggestion
// alternative is to store the indices in a separate array (called micro column layout), which is desirable when this
```

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -520,9 +524,82 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+// Tuning it too high might invalidate the gains from `reduce_tlb_pressure`.
+constexpr auto MIN_CACHELINES_PER_STORE = size_t{1};
+
+// TemporaryRadixBucket is one "chunk" of data that can be copied to one specific partition once full. It is
```
Can you add half a sentence about the purpose? It's tuples of a larger radix partition, but we don't want to write every single item to each partition but gather multiple in a bucket and write them alltogether.

Something like that.

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -520,9 +524,82 @@ std::vector<std::optional<PosHashTable<HashedType>>> build(const RadixContainer<
+  // We want the size of this container to be a multiple of the element size (so we don't copy uninitialized bytes) and
+  // the cacheline size (so we can use aligned vector instructions to copy). Additionally, we want the size of this
+  // container to be at least MIN_CACHELINES_PER_STORE cachelines large. If BYTES_PER_ELEMENT is not a power of two,
+  // we have to extract the part that is not a power of 2 and use it as a factor. For all types in Hyrise, this is
```
`If BYTES_PER_ELEMENT is not a power of two, we have to extract the part that is not a power of 2 and use it as a factor.`

Can you rewrite this part? Or give an example. You list that this might be 1 (but 8 is a power of two?) or 3, but how does that relate to "the factor" and how do you come up with 3?

Greenscreen23 replied:
Better?

Bouncner commented:
Can you please submit the plotting script?

Greenscreen23 commented:
Some more food for thought: If I try to list the instantiations of `partition_by_radix` in the `libhyrise_impl.so`, I only find variant N = 0 and locality = 0, which is never called in the benchmark. The benchmark either calls variant = A|B|C|D and locality = 0 or variant = N and locality = -1. This implies that these are not in the hyrise lib, but are currently instantiated in the benchmark binary. This should not be much of an issue, as all of them are instantiated there. Are there some performance optimizations applied to the hyrise library that are not applied to the benchmark binary?

I used this command to find this out
```
llvm-objdump-20 -t -C ./hyriseMicroBenchmarks | grep "partition_by_radix"
llvm-objdump-20 -t -C ./lib/libhyrise_impl.so | grep "partition_by_radix"
```

Greenscreen23 commented:
Hey, I have done some single threaded perf measurements. Maybe those are helpful to you @Bouncner? All of them were measured on nemea for 100 benchmark iterations with the performance cpu governor. Each partition had 1e5 items and there were 2^17 partitions. For the variants with prefetch instructions, the locality was always 0 (non temporal).

These are (some of the) the results for `perf stat -d -d -d`:
- The number of branches in A and B is higher, which is due to the branch required when prefetching. The branch miss percentage is lower, but the absolute number is higher.
- Variant D has a higher L1-dcache-load-miss percentage, which might indicate that always prefetching the next cacheline causes other useful entries to be evicted from the L1-dcache.
- Variant C has more L1-dcache-loads, which might be due to the more aggressive prefetching.
- Variant C and N have the highest LLC (last level cache) load count while also having the lowest miss percentage. This means they are probably better at loading the correct things

In general it seems like C and N are still the best options, which is similar to our previous experiments.

| Variant | instructions (per cycle) | branches | branch-misses %| L1-dcache-loads | L1-dcache-load-misses % | LLC-loads | LLC-load-misses % | dTLB-loads | dTLB-load-misses % | iTLB-loads | iTLB-load-misses % |
| - | - | - | - | - | - | - | - | - | - | - | - |
| A | 0.69 | 370.206 M/sec | 0.13 | 766.108 M/sec | 17.45 | 11.675 M/sec | 15.21 | 766.634 M/sec | 0.25 | 681.502 K/sec | 20.79 |
| B | 0.68 | 367.928 M/sec | 0.13 | 763.461 M/sec | 17.82 | 11.920 M/sec | 17.82 | 765.841 M/sec | 0.25 | 661.341 K/sec | 20.95 |
| C | 0.65 | 340.309 M/sec | 0.14 | 804.364 M/sec | 17.20 | 14.499 M/sec | 13.14 | 804.535 M/sec | 0.23 | 648.726 K/sec | 21.05 |
| D | 0.64 | 335.405 M/sec | 0.14 | 786.481 M/sec | 19.12 | 15.612 M/sec | 16.15 | 786.661 M/sec | 0.23 | 656.957 K/sec | 20.96 | 
| N | 0.64 | 340.013 M/sec | 0.14 | 766.194 M/sec | 17.70 | 15.518 M/sec | 13.14 | 765.963 M/sec | 0.25 | 653.695 K/sec | 21.27 | 

I also ran `perf stat --topdown`, but I can't extract any useful information out of these values. I am also not sure if I mapped the % signs correctly to the columns. The perf output was kind of garbled. However, with 4 percentage signs and one value that does not seem to be a percentage, I felt like this allocation is correct.

| Variant | tma_info_core_ipmispredict | tma_frontend_bound % | tma_retiring % | tma_backend_bound % | tma_bad_speculation % |
| - | - | - | - | - | - |
| A |5320.71 | 10.8 | 18.2| 69.8| 1.2 |
| B | 5386.30 | 10.4 | 18.1| 70.2| 1.2| 
| C | 5072.20 | 10.6 | 17.4 | 70.8 | 1.3 |
| D | 4369.75 | 12.5 | 18.0 | 68.1 | 1.4 |
| N | 5079.84 | 10.5 | 17.0 | 71.2 | 1.2 |

Greenscreen23 commented:
Here is the assembly of the hot loop prefetching of all variants. The source was compiled on nemea in cmake release mode (which implies `-O3`) with `-g -fno-omit-frame-pointer`. The functions were all created in the `hyriseMicroBenchmark` binary, as mentioned above.

Variants A and B generate quite a bit of code to do the comparison, including a jump. C and D can just use a single instruction, which probably gets less in the way.

Variant A:
```
;             if ((std::bit_cast<uint64_t>(tmp.data[radix].elements.data() + tmp_idx + 1) & (BYTES_PER_CACHELINE - 1)) <
   9b79f: 4f 8d 1c ba                  	lea	r11, [r10 + 4*r15]
   9b7a3: 49 83 c3 0c                  	add	r11, 0xc
   9b7a7: 45 89 df                     	mov	r15d, r11d
   9b7aa: 41 83 e7 3c                  	and	r15d, 0x3c
   9b7ae: 41 83 ff 0b                  	cmp	r15d, 0xb
   9b7b2: 77 04                        	ja	0x9b7b8 <_ZZN6hyrise18partition_by_radixIjmLb0ELb1ELc65ELi0EEESt6vectorINS_9PartitionIT_EESaIS4_EERKS6_RKS1_IS1_ImSaImEESaISA_EEmRKN5boost14dynamic_bitsetImS9_EEENKUlvE_clEv+0x198>
;               __builtin_prefetch(tmp.data[radix].elements.data() + tmp_idx + 1, 1, locality);
   9b7b4: 41 0f 0d 0b                  	prefetchw	byte ptr [r11]
```
Variant B:
```
;             if ((std::bit_cast<uint64_t>(reinterpret_cast<std::byte*>(tmp.data[radix].elements.data() + tmp_idx + 2) -
   9be33: 45 8d 5f 17                  	lea	r11d, [r15 + 0x17]
;                                          1) &
   9be37: 41 83 e3 3c                  	and	r11d, 0x3c
;                  (BYTES_PER_CACHELINE - 1)) < sizeof(PartitionedElement<T>)) {
   9be3b: 41 83 fb 0b                  	cmp	r11d, 0xb
;             if ((std::bit_cast<uint64_t>(reinterpret_cast<std::byte*>(tmp.data[radix].elements.data() + tmp_idx + 2) -
   9be3f: 77 05                        	ja	0x9be46 <_ZZN6hyrise18partition_by_radixIjmLb0ELb1ELc66ELi0EEESt6vectorINS_9PartitionIT_EESaIS4_EERKS6_RKS1_IS1_ImSaImEESaISA_EEmRKN5boost14dynamic_bitsetImS9_EEENKUlvE_clEv+0x196>
;               __builtin_prefetch(tmp.data[radix].elements.data() + tmp_idx + 2, 1, locality);
   9be41: 41 0f 0d 4f 18               	prefetchw	byte ptr [r15 + 0x18]
```
Variant C:
```
;             __builtin_prefetch(tmp.data[radix].elements.data() + tmp_idx + 1, 1, locality);
   9c295: 43 0f 0d 4c bb 0c            	prefetchw	byte ptr [r11 + 4*r15 + 0xc]
```
Variant D:
```
;             __builtin_prefetch(
   9c6e5: 43 0f 0d 4c bb 40            	prefetchw	byte ptr [r11 + 4*r15 + 0x40]
```

Greenscreen23 commented:
I could try to benchmark the functions with `perf record` or even `perf mem record` to see how bad this jump really is. What do you think @Bouncner?

Bouncner commented:
> I could try to benchmark the functions with `perf record` or even `perf mem record` to see how bad this jump really is. What do you think @Bouncner?

Thanks a lot for all the input!

I am currently running some benchmarks on my own. I first need to see what the results are as I have changed a few things (multithreading mostly).

As I am not fully sure what the impact of the different locations is, I changed the code to have instantiations in libHyrise for all variants. I probably need to do the perf measurements again when I have a better idea about the MT performance.

Greenscreen23 commented:
> I am currently running some benchmarks on my own. I first need to see what the results are as I have changed a few things (multithreading mostly).

I compared your branch with mine, but I only saw your MT changes to the benchmark and the instantiation of the function. These should not invalidate my results, right?

Are you sure that your way of instantiating the function actually results in calls to the library for that function? The function is fully implemented in a header, so any include of that header will generate another instance in the binary, will it not?

Is it even bad that the implementation is generated in the binary? It should be compiled with the same flags, so it should have the same assembly, or not?

> As I am not fully sure what the impact of the different locations is, I changed the code to have instantiations in libHyrise for all variants. I probably need to do the perf measurements again when I have a better idea about the MT performance.

What do you expect when you measure the MT performance? I could image that LLC contention and TLB pressure increase but it should not do affect L1? Maybe the performance of A B is better, because they are more selective with prefetching? I could also image that A and B are still slower, if the branch does more harm than the prefetch does good.

Bouncner commented:
> I compared your branch with mine, but I only saw your MT changes to the benchmark and the instantiation of the function. These should not invalidate my results, right?

It's not all committed. Still playing around. One thing I was not sure about is if there are patterns in the data (as we added `value++`) that lead to different results than a more typical "random-ish" pattern. I am thinking about fixed strides to write and such things.
 
> Are you sure that your way of instantiating the function actually results in calls to the library for that function? The function is fully implemented in a header, so any include of that header will generate another instance in the binary, will it not?

The linker should resolve this.
Checking the binary also looks like all implementations are in the  library.

> Is it even bad that the implementation is generated in the binary? It should be compiled with the same flags, so it should have the same assembly, or not?

I compiled without LTO. So I can rule out that. But I am not sure what the effects on LTO would be.
Generally, the compiler could make different decisions based on the context of the function (inline or not). I wouldn't be surprised if the compiler generates different code. But to be honest, I have no idea.

> What do you expect when you measure the MT performance? I could image that LLC contention and TLB pressure increase but it should not do affect L1? Maybe the performance of A B is better, because they are more selective with prefetching? I could also image that A and B are still slower, if the branch does more harm than the prefetch does good.

TLB pressure should increase, e.g., with hyper-threading (shared TLB cache). So there is a difference. Concerning the ABCDN measurements on prefetching: we generate much more outstanding loads. I want to be sure we measure something reflecting the actual workload when running TPC-H/JOB etc.

Greenscreen23 commented:
Some more food for thought: There does not seem to be a difference between all of the locality variants (0 for NTA and 1-3 for different levels). There are instructions for read prefetching for each of these, but write prefetching always seems to be `prefetchw`. I compiled this test program with clang and gcc with `-O3 -march=native` on nemea and the hpc machines
```cpp
#include <bits/stdc++.h>
using namespace std;

template <int locality> void prefetch(int argc, char **argv) {
  for (auto idx = int{0}; idx < argc; idx++) {
    __builtin_prefetch(argv[idx], 1, locality);
  }
}

template void prefetch<0>(int argc, char **argv);
template void prefetch<1>(int argc, char **argv);
template void prefetch<2>(int argc, char **argv);
template void prefetch<3>(int argc, char **argv);

int main() {
    return 0;
}
```

clang AMD
```asm
0000000000001830 <void prefetch<0>(int, char**)>:
    1830: 85 ff                         test    edi, edi
    1832: 7e 1b                         jle     0x184f <void prefetch<0>(int, char**)+0x1f>
    1834: 89 f8                         mov     eax, edi
    1836: 31 c9                         xor     ecx, ecx
    1838: 0f 1f 84 00 00 00 00 00       nop     dword ptr [rax + rax]
    1840: 48 8b 14 ce                   mov     rdx, qword ptr [rsi + 8*rcx]
    1844: 48 ff c1                      inc     rcx
    1847: 0f 0d 0a                      prefetchw       byte ptr [rdx]
    184a: 48 39 c8                      cmp     rax, rcx
    184d: 75 f1                         jne     0x1840 <void prefetch<0>(int, char**)+0x10>
    184f: c3                            ret

0000000000001850 <void prefetch<1>(int, char**)>:
    1850: 85 ff                         test    edi, edi
    1852: 7e 1b                         jle     0x186f <void prefetch<1>(int, char**)+0x1f>
    1854: 89 f8                         mov     eax, edi
    1856: 31 c9                         xor     ecx, ecx
    1858: 0f 1f 84 00 00 00 00 00       nop     dword ptr [rax + rax]
    1860: 48 8b 14 ce                   mov     rdx, qword ptr [rsi + 8*rcx]
    1864: 48 ff c1                      inc     rcx
    1867: 0f 0d 0a                      prefetchw       byte ptr [rdx]
    186a: 48 39 c8                      cmp     rax, rcx
    186d: 75 f1                         jne     0x1860 <void prefetch<1>(int, char**)+0x10>
    186f: c3                            ret

0000000000001870 <void prefetch<2>(int, char**)>:
    1870: 85 ff                         test    edi, edi
    1872: 7e 1b                         jle     0x188f <void prefetch<2>(int, char**)+0x1f>
    1874: 89 f8                         mov     eax, edi
    1876: 31 c9                         xor     ecx, ecx
    1878: 0f 1f 84 00 00 00 00 00       nop     dword ptr [rax + rax]
    1880: 48 8b 14 ce                   mov     rdx, qword ptr [rsi + 8*rcx]
    1884: 48 ff c1                      inc     rcx
    1887: 0f 0d 0a                      prefetchw       byte ptr [rdx]
    188a: 48 39 c8                      cmp     rax, rcx
    188d: 75 f1                         jne     0x1880 <void prefetch<2>(int, char**)+0x10>
    188f: c3                            ret

0000000000001890 <void prefetch<3>(int, char**)>:
    1890: 85 ff                         test    edi, edi
    1892: 7e 1b                         jle     0x18af <void prefetch<3>(int, char**)+0x1f>
    1894: 89 f8                         mov     eax, edi
    1896: 31 c9                         xor     ecx, ecx
    1898: 0f 1f 84 00 00 00 00 00       nop     dword ptr [rax + rax]
    18a0: 48 8b 14 ce                   mov     rdx, qword ptr [rsi + 8*rcx]
    18a4: 48 ff c1                      inc     rcx
    18a7: 0f 0d 0a                      prefetchw       byte ptr [rdx]
    18aa: 48 39 c8                      cmp     rax, rcx
    18ad: 75 f1                         jne     0x18a0 <void prefetch<3>(int, char**)+0x10>
    18af: c3                            ret
```
gcc AMD
```asm
0000000000001890 <void prefetch<0>(int, char**)>:
    1890: f3 0f 1e fa                   endbr64
    1894: 85 ff                         test    edi, edi
    1896: 7e 17                         jle     0x18af <void prefetch<0>(int, char**)+0x1f>
    1898: 48 63 ff                      movsxd  rdi, edi
    189b: 48 8d 14 fe                   lea     rdx, [rsi + 8*rdi]
    189f: 90                            nop
    18a0: 48 8b 06                      mov     rax, qword ptr [rsi]
    18a3: 48 83 c6 08                   add     rsi, 0x8
    18a7: 0f 0d 08                      prefetchw       byte ptr [rax]
    18aa: 48 39 d6                      cmp     rsi, rdx
    18ad: 75 f1                         jne     0x18a0 <void prefetch<0>(int, char**)+0x10>
    18af: c3                            ret

00000000000018b0 <void prefetch<1>(int, char**)>:
    18b0: f3 0f 1e fa                   endbr64
    18b4: 85 ff                         test    edi, edi
    18b6: 7e 17                         jle     0x18cf <void prefetch<1>(int, char**)+0x1f>
    18b8: 48 63 ff                      movsxd  rdi, edi
    18bb: 48 8d 14 fe                   lea     rdx, [rsi + 8*rdi]
    18bf: 90                            nop
    18c0: 48 8b 06                      mov     rax, qword ptr [rsi]
    18c3: 48 83 c6 08                   add     rsi, 0x8
    18c7: 0f 0d 08                      prefetchw       byte ptr [rax]
    18ca: 48 39 d6                      cmp     rsi, rdx
    18cd: 75 f1                         jne     0x18c0 <void prefetch<1>(int, char**)+0x10>
    18cf: c3                            ret

00000000000018d0 <void prefetch<2>(int, char**)>:
    18d0: f3 0f 1e fa                   endbr64
    18d4: 85 ff                         test    edi, edi
    18d6: 7e 17                         jle     0x18ef <void prefetch<2>(int, char**)+0x1f>
    18d8: 48 63 ff                      movsxd  rdi, edi
    18db: 48 8d 14 fe                   lea     rdx, [rsi + 8*rdi]
    18df: 90                            nop
    18e0: 48 8b 06                      mov     rax, qword ptr [rsi]
    18e3: 48 83 c6 08                   add     rsi, 0x8
    18e7: 0f 0d 08                      prefetchw       byte ptr [rax]
    18ea: 48 39 d6                      cmp     rsi, rdx
    18ed: 75 f1                         jne     0x18e0 <void prefetch<2>(int, char**)+0x10>
    18ef: c3                            ret

00000000000018f0 <void prefetch<3>(int, char**)>:
    18f0: f3 0f 1e fa                   endbr64
    18f4: 85 ff                         test    edi, edi
    18f6: 7e 17                         jle     0x190f <void prefetch<3>(int, char**)+0x1f>
    18f8: 48 63 ff                      movsxd  rdi, edi
    18fb: 48 8d 14 fe                   lea     rdx, [rsi + 8*rdi]
    18ff: 90                            nop
    1900: 48 8b 06                      mov     rax, qword ptr [rsi]
    1903: 48 83 c6 08                   add     rsi, 0x8
    1907: 0f 0d 08                      prefetchw       byte ptr [rax]
    190a: 48 39 d6                      cmp     rsi, rdx
    190d: 75 f1                         jne     0x1900 <void prefetch<3>(int, char**)+0x10>
    190f: c3                            ret
```
clang INTEL
```asm
00000000000017b0 <void prefetch<0>(int, char**)>:
    17b0: 85 ff                         test    edi, edi
    17b2: 0f 8e 87 00 00 00             jle     0x183f <void prefetch<0>(int, char**)+0x8f>
    17b8: 89 f9                         mov     ecx, edi
    17ba: 89 c8                         mov     eax, ecx
    17bc: 83 e0 07                      and     eax, 0x7
    17bf: 83 ff 08                      cmp     edi, 0x8
    17c2: 73 04                         jae     0x17c8 <void prefetch<0>(int, char**)+0x18>
    17c4: 31 d2                         xor     edx, edx
    17c6: eb 50                         jmp     0x1818 <void prefetch<0>(int, char**)+0x68>
    17c8: 81 e1 f8 ff ff 7f             and     ecx, 0x7ffffff8
    17ce: 31 d2                         xor     edx, edx
    17d0: 48 8b 3c d6                   mov     rdi, qword ptr [rsi + 8*rdx]
    17d4: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    17d7: 48 8b 7c d6 08                mov     rdi, qword ptr [rsi + 8*rdx + 0x8]
    17dc: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    17df: 48 8b 7c d6 10                mov     rdi, qword ptr [rsi + 8*rdx + 0x10]
    17e4: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    17e7: 48 8b 7c d6 18                mov     rdi, qword ptr [rsi + 8*rdx + 0x18]
    17ec: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    17ef: 48 8b 7c d6 20                mov     rdi, qword ptr [rsi + 8*rdx + 0x20]
    17f4: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    17f7: 48 8b 7c d6 28                mov     rdi, qword ptr [rsi + 8*rdx + 0x28]
    17fc: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    17ff: 48 8b 7c d6 30                mov     rdi, qword ptr [rsi + 8*rdx + 0x30]
    1804: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1807: 48 8b 7c d6 38                mov     rdi, qword ptr [rsi + 8*rdx + 0x38]
    180c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    180f: 48 83 c2 08                   add     rdx, 0x8
    1813: 48 39 d1                      cmp     rcx, rdx
    1816: 75 b8                         jne     0x17d0 <void prefetch<0>(int, char**)+0x20>
    1818: 48 85 c0                      test    rax, rax
    181b: 74 22                         je      0x183f <void prefetch<0>(int, char**)+0x8f>
    181d: 48 8d 0c d6                   lea     rcx, [rsi + 8*rdx]
    1821: 31 d2                         xor     edx, edx
    1823: 66 66 66 66 2e 0f 1f 84 00 00 00 00 00        nop     word ptr cs:[rax + rax]
    1830: 48 8b 34 d1                   mov     rsi, qword ptr [rcx + 8*rdx]
    1834: 0f 0d 0e                      prefetchw       byte ptr [rsi]
    1837: 48 ff c2                      inc     rdx
    183a: 48 39 d0                      cmp     rax, rdx
    183d: 75 f1                         jne     0x1830 <void prefetch<0>(int, char**)+0x80>
    183f: c3                            ret

0000000000001840 <void prefetch<1>(int, char**)>:
    1840: 85 ff                         test    edi, edi
    1842: 0f 8e 87 00 00 00             jle     0x18cf <void prefetch<1>(int, char**)+0x8f>
    1848: 89 f9                         mov     ecx, edi
    184a: 89 c8                         mov     eax, ecx
    184c: 83 e0 07                      and     eax, 0x7
    184f: 83 ff 08                      cmp     edi, 0x8
    1852: 73 04                         jae     0x1858 <void prefetch<1>(int, char**)+0x18>
    1854: 31 d2                         xor     edx, edx
    1856: eb 50                         jmp     0x18a8 <void prefetch<1>(int, char**)+0x68>
    1858: 81 e1 f8 ff ff 7f             and     ecx, 0x7ffffff8
    185e: 31 d2                         xor     edx, edx
    1860: 48 8b 3c d6                   mov     rdi, qword ptr [rsi + 8*rdx]
    1864: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1867: 48 8b 7c d6 08                mov     rdi, qword ptr [rsi + 8*rdx + 0x8]
    186c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    186f: 48 8b 7c d6 10                mov     rdi, qword ptr [rsi + 8*rdx + 0x10]
    1874: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1877: 48 8b 7c d6 18                mov     rdi, qword ptr [rsi + 8*rdx + 0x18]
    187c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    187f: 48 8b 7c d6 20                mov     rdi, qword ptr [rsi + 8*rdx + 0x20]
    1884: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1887: 48 8b 7c d6 28                mov     rdi, qword ptr [rsi + 8*rdx + 0x28]
    188c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    188f: 48 8b 7c d6 30                mov     rdi, qword ptr [rsi + 8*rdx + 0x30]
    1894: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1897: 48 8b 7c d6 38                mov     rdi, qword ptr [rsi + 8*rdx + 0x38]
    189c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    189f: 48 83 c2 08                   add     rdx, 0x8
    18a3: 48 39 d1                      cmp     rcx, rdx
    18a6: 75 b8                         jne     0x1860 <void prefetch<1>(int, char**)+0x20>
    18a8: 48 85 c0                      test    rax, rax
    18ab: 74 22                         je      0x18cf <void prefetch<1>(int, char**)+0x8f>
    18ad: 48 8d 0c d6                   lea     rcx, [rsi + 8*rdx]
    18b1: 31 d2                         xor     edx, edx
    18b3: 66 66 66 66 2e 0f 1f 84 00 00 00 00 00        nop     word ptr cs:[rax + rax]
    18c0: 48 8b 34 d1                   mov     rsi, qword ptr [rcx + 8*rdx]
    18c4: 0f 0d 0e                      prefetchw       byte ptr [rsi]
    18c7: 48 ff c2                      inc     rdx
    18ca: 48 39 d0                      cmp     rax, rdx
    18cd: 75 f1                         jne     0x18c0 <void prefetch<1>(int, char**)+0x80>
    18cf: c3                            ret

00000000000018d0 <void prefetch<2>(int, char**)>:
    18d0: 85 ff                         test    edi, edi
    18d2: 0f 8e 87 00 00 00             jle     0x195f <void prefetch<2>(int, char**)+0x8f>
    18d8: 89 f9                         mov     ecx, edi
    18da: 89 c8                         mov     eax, ecx
    18dc: 83 e0 07                      and     eax, 0x7
    18df: 83 ff 08                      cmp     edi, 0x8
    18e2: 73 04                         jae     0x18e8 <void prefetch<2>(int, char**)+0x18>
    18e4: 31 d2                         xor     edx, edx
    18e6: eb 50                         jmp     0x1938 <void prefetch<2>(int, char**)+0x68>
    18e8: 81 e1 f8 ff ff 7f             and     ecx, 0x7ffffff8
    18ee: 31 d2                         xor     edx, edx
    18f0: 48 8b 3c d6                   mov     rdi, qword ptr [rsi + 8*rdx]
    18f4: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    18f7: 48 8b 7c d6 08                mov     rdi, qword ptr [rsi + 8*rdx + 0x8]
    18fc: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    18ff: 48 8b 7c d6 10                mov     rdi, qword ptr [rsi + 8*rdx + 0x10]
    1904: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1907: 48 8b 7c d6 18                mov     rdi, qword ptr [rsi + 8*rdx + 0x18]
    190c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    190f: 48 8b 7c d6 20                mov     rdi, qword ptr [rsi + 8*rdx + 0x20]
    1914: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1917: 48 8b 7c d6 28                mov     rdi, qword ptr [rsi + 8*rdx + 0x28]
    191c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    191f: 48 8b 7c d6 30                mov     rdi, qword ptr [rsi + 8*rdx + 0x30]
    1924: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1927: 48 8b 7c d6 38                mov     rdi, qword ptr [rsi + 8*rdx + 0x38]
    192c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    192f: 48 83 c2 08                   add     rdx, 0x8
    1933: 48 39 d1                      cmp     rcx, rdx
    1936: 75 b8                         jne     0x18f0 <void prefetch<2>(int, char**)+0x20>
    1938: 48 85 c0                      test    rax, rax
    193b: 74 22                         je      0x195f <void prefetch<2>(int, char**)+0x8f>
    193d: 48 8d 0c d6                   lea     rcx, [rsi + 8*rdx]
    1941: 31 d2                         xor     edx, edx
    1943: 66 66 66 66 2e 0f 1f 84 00 00 00 00 00        nop     word ptr cs:[rax + rax]
    1950: 48 8b 34 d1                   mov     rsi, qword ptr [rcx + 8*rdx]
    1954: 0f 0d 0e                      prefetchw       byte ptr [rsi]
    1957: 48 ff c2                      inc     rdx
    195a: 48 39 d0                      cmp     rax, rdx
    195d: 75 f1                         jne     0x1950 <void prefetch<2>(int, char**)+0x80>
    195f: c3                            ret

0000000000001960 <void prefetch<3>(int, char**)>:
    1960: 85 ff                         test    edi, edi
    1962: 0f 8e 87 00 00 00             jle     0x19ef <void prefetch<3>(int, char**)+0x8f>
    1968: 89 f9                         mov     ecx, edi
    196a: 89 c8                         mov     eax, ecx
    196c: 83 e0 07                      and     eax, 0x7
    196f: 83 ff 08                      cmp     edi, 0x8
    1972: 73 04                         jae     0x1978 <void prefetch<3>(int, char**)+0x18>
    1974: 31 d2                         xor     edx, edx
    1976: eb 50                         jmp     0x19c8 <void prefetch<3>(int, char**)+0x68>
    1978: 81 e1 f8 ff ff 7f             and     ecx, 0x7ffffff8
    197e: 31 d2                         xor     edx, edx
    1980: 48 8b 3c d6                   mov     rdi, qword ptr [rsi + 8*rdx]
    1984: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1987: 48 8b 7c d6 08                mov     rdi, qword ptr [rsi + 8*rdx + 0x8]
    198c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    198f: 48 8b 7c d6 10                mov     rdi, qword ptr [rsi + 8*rdx + 0x10]
    1994: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    1997: 48 8b 7c d6 18                mov     rdi, qword ptr [rsi + 8*rdx + 0x18]
    199c: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    199f: 48 8b 7c d6 20                mov     rdi, qword ptr [rsi + 8*rdx + 0x20]
    19a4: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    19a7: 48 8b 7c d6 28                mov     rdi, qword ptr [rsi + 8*rdx + 0x28]
    19ac: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    19af: 48 8b 7c d6 30                mov     rdi, qword ptr [rsi + 8*rdx + 0x30]
    19b4: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    19b7: 48 8b 7c d6 38                mov     rdi, qword ptr [rsi + 8*rdx + 0x38]
    19bc: 0f 0d 0f                      prefetchw       byte ptr [rdi]
    19bf: 48 83 c2 08                   add     rdx, 0x8
    19c3: 48 39 d1                      cmp     rcx, rdx
    19c6: 75 b8                         jne     0x1980 <void prefetch<3>(int, char**)+0x20>
    19c8: 48 85 c0                      test    rax, rax
    19cb: 74 22                         je      0x19ef <void prefetch<3>(int, char**)+0x8f>
    19cd: 48 8d 0c d6                   lea     rcx, [rsi + 8*rdx]
    19d1: 31 d2                         xor     edx, edx
    19d3: 66 66 66 66 2e 0f 1f 84 00 00 00 00 00        nop     word ptr cs:[rax + rax]
    19e0: 48 8b 34 d1                   mov     rsi, qword ptr [rcx + 8*rdx]
    19e4: 0f 0d 0e                      prefetchw       byte ptr [rsi]
    19e7: 48 ff c2                      inc     rdx
    19ea: 48 39 d0                      cmp     rax, rdx
    19ed: 75 f1                         jne     0x19e0 <void prefetch<3>(int, char**)+0x80>
    19ef: c3                            ret
```
gcc INTEL
```asm
00000000000011c0 <void prefetch<0>(int, char**)>:
    11c0: f3 0f 1e fa                   endbr64
    11c4: 85 ff                         test    edi, edi
    11c6: 7e 17                         jle     0x11df <void prefetch<0>(int, char**)+0x1f>
    11c8: 8d 47 ff                      lea     eax, [rdi - 0x1]
    11cb: 48 8d 54 c6 08                lea     rdx, [rsi + 8*rax + 0x8]
    11d0: 48 8b 06                      mov     rax, qword ptr [rsi]
    11d3: 48 83 c6 08                   add     rsi, 0x8
    11d7: 0f 0d 08                      prefetchw       byte ptr [rax]
    11da: 48 39 d6                      cmp     rsi, rdx
    11dd: 75 f1                         jne     0x11d0 <void prefetch<0>(int, char**)+0x10>
    11df: c3                            ret

00000000000011e0 <void prefetch<1>(int, char**)>:
    11e0: f3 0f 1e fa                   endbr64
    11e4: 85 ff                         test    edi, edi
    11e6: 7e 17                         jle     0x11ff <void prefetch<1>(int, char**)+0x1f>
    11e8: 8d 47 ff                      lea     eax, [rdi - 0x1]
    11eb: 48 8d 54 c6 08                lea     rdx, [rsi + 8*rax + 0x8]
    11f0: 48 8b 06                      mov     rax, qword ptr [rsi]
    11f3: 48 83 c6 08                   add     rsi, 0x8
    11f7: 0f 0d 08                      prefetchw       byte ptr [rax]
    11fa: 48 39 d6                      cmp     rsi, rdx
    11fd: 75 f1                         jne     0x11f0 <void prefetch<1>(int, char**)+0x10>
    11ff: c3                            ret

0000000000001200 <void prefetch<2>(int, char**)>:
    1200: f3 0f 1e fa                   endbr64
    1204: 85 ff                         test    edi, edi
    1206: 7e 17                         jle     0x121f <void prefetch<2>(int, char**)+0x1f>
    1208: 8d 47 ff                      lea     eax, [rdi - 0x1]
    120b: 48 8d 54 c6 08                lea     rdx, [rsi + 8*rax + 0x8]
    1210: 48 8b 06                      mov     rax, qword ptr [rsi]
    1213: 48 83 c6 08                   add     rsi, 0x8
    1217: 0f 0d 08                      prefetchw       byte ptr [rax]
    121a: 48 39 d6                      cmp     rsi, rdx
    121d: 75 f1                         jne     0x1210 <void prefetch<2>(int, char**)+0x10>
    121f: c3                            ret

0000000000001220 <void prefetch<3>(int, char**)>:
    1220: f3 0f 1e fa                   endbr64
    1224: 85 ff                         test    edi, edi
    1226: 7e 17                         jle     0x123f <void prefetch<3>(int, char**)+0x1f>
    1228: 8d 47 ff                      lea     eax, [rdi - 0x1]
    122b: 48 8d 54 c6 08                lea     rdx, [rsi + 8*rax + 0x8]
    1230: 48 8b 06                      mov     rax, qword ptr [rsi]
    1233: 48 83 c6 08                   add     rsi, 0x8
    1237: 0f 0d 08                      prefetchw       byte ptr [rax]
    123a: 48 39 d6                      cmp     rsi, rdx
    123d: 75 f1                         jne     0x1230 <void prefetch<3>(int, char**)+0x10>
    123f: c3                            ret
```

Greenscreen23 commented:
> > I compared your branch with mine, but I only saw your MT changes to the benchmark and the instantiation of the function. These should not invalidate my results, right?
> 
> It's not all committed. Still playing around. One thing I was not sure about is if there are patterns in the data (as we added `value++`) that lead to different results than a more typical "random-ish" pattern. I am thinking about fixed strides to write and such things.

Hmm, I do agree that this could be interesting. But I think you should adapt the distribution of data rather than the pattern. My current setup should cause the next cache line to never be in the cache, as all the previous ones are in the cache. I think a simple stride just leads to a different order that destinations are written to without changing the situation. With a different distribution, one might really see the prefetching advantage for buckets occuring more often.

> > Are you sure that your way of instantiating the function actually results in calls to the library for that function? The function is fully implemented in a header, so any include of that header will generate another instance in the binary, will it not?
> 
> The linker should resolve this. Checking the binary also looks like all implementations are in the library.

Huh, I did not know the linker does such things. But if it really is not in the binary, then it surely did.

> > What do you expect when you measure the MT performance? I could image that LLC contention and TLB pressure increase but it should not do affect L1? Maybe the performance of A B is better, because they are more selective with prefetching? I could also image that A and B are still slower, if the branch does more harm than the prefetch does good.
> 
> TLB pressure should increase, e.g., with hyper-threading (shared TLB cache). So there is a difference. Concerning the ABCDN measurements on prefetching: we generate much more outstanding loads. I want to be sure we measure something reflecting the actual workload when running TPC-H/JOB etc.

What is a more outstanding load? Just more elements? A different distribution? Other operators running in the background?

Bouncner commented:
> Hmm, I do agree that this could be interesting. But I think you should adapt the distribution of data rather than the pattern. My current setup should cause the next cache line to never be in the cache, as all the previous ones are in the cache. I think a simple stride just leads to a different order that destinations are written to without changing the situation. With a different distribution, one might really see the prefetching advantage for buckets occuring more often.

To be honest, I am not focusing non-uniform distributions as it makes everything much harder to analyze and our typical benchmarks are mostly unskewed.

> What is a more outstanding load? Just more elements? A different distribution? Other operators running in the background?

I meant outstanding prefetching requests. Marcel is our expert here, but the number of requests you can be handled by the CPU is limited. But I just took a look into the most recent paper from Marcel in this regard (https://hpi.de/oldsite/fileadmin/user_upload/fachgebiete/rabl/publications/2025/Mahling-DaMoN25-Prefetching.pdf) and it reads as if the requests are per core. In this case, we probably never hit this limit and MT does not change anything here.

MT already brought one surprising insight: for many partitions, the actual radix partitioning is not the bottleneck. It's the calculation of the offsets. And that's ST.
I need to wrap my head around that.

Bouncner commented:
I think I need to think a bit more about the benchmarks. What I totally neglected was our scheduling.

Let's say we have 2^14 partitions. We have thousands of radix buckets (just obtaining the offset to write from the histogram aggregation is easily out of L1 cache). As we schedule per chunk, we end up with just a few 2/4/8 items per radix partition. No surprise we do not see any changes in the large radix bits.

I am running a few benchmarks (when I get a slot at the SCI) to see how much time we spent in the radix partitioning over all (TPC-H). If it's a lot (especially with your probe parallelization) we need to rewrite the radix partitioning.

Greenscreen23 commented:
> > What is a more outstanding load? Just more elements? A different distribution? Other operators running in the background?
> 
> I meant outstanding prefetching requests. Marcel is our expert here, but the number of requests you can be handled by the CPU is limited. But I just took a look into the most recent paper from Marcel in this regard (https://hpi.de/oldsite/fileadmin/user_upload/fachgebiete/rabl/publications/2025/Mahling-DaMoN25-Prefetching.pdf) and it reads as if the requests are per core. In this case, we probably never hit this limit and MT does not change anything here.

The paper says that this "Fill Buffer" (where the prefetch requests go) is located between L1 and L2 cache, which are both per core on our cpus. I think we use hyperthreading, right? In this case, FB could be shared between the two threads on a core. That's only a factor of two, but without numbers on the size of the FB, it is hard to tell whether we are already overusing it or if it's empty most of the time.

> MT already brought one surprising insight: for many partitions, the actual radix partitioning is not the bottleneck. It's the calculation of the offsets. And that's ST. I need to wrap my head around that.

You are talking about calculating `output_offsets_by_input_partition`, right? We could use prefetching here, as we know where which part of the nested vector struct we want to access next, but the compiler probably does not. This part should also be parallelizable. What do you think of this code snippet?

```cpp
auto jobs = std::vector<std::shared_ptr<AbstractTask>>{};
jobs.reserve(output_partition_count);

// output_offsets_by_input_partition[input_partition_idx][output_partition_idx] holds the first offset in the
// bucket written for input_partition_idx
auto output_offsets_by_input_partition =
    std::vector(input_partition_count, std::vector<size_t>(output_partition_count));
for (auto output_partition_idx = size_t{0}; output_partition_idx < output_partition_count; ++output_partition_idx) {
  const auto calculate_offsets = [&, output_partition_idx]() {
    auto this_output_partition_size = size_t{0};
    for (auto input_partition_idx = size_t{0}; input_partition_idx < input_partition_count; ++input_partition_idx) {
      // constexpr auto STRIDE = size_t{3};
      // if (input_partition_idx + STRIDE < input_partition_count) {
      //   __builtin_prefetch(output_offsets_by_input_partition[input_partition_idx + STRIDE].data() + output_partition_idx, 1,
      //                      0);
      //   __builtin_prefetch(histograms[input_partition_idx + STRIDE].data() + output_partition_idx, 0, 0);
      // }
      output_offsets_by_input_partition[input_partition_idx][output_partition_idx] = this_output_partition_size;
      this_output_partition_size += histograms[input_partition_idx][output_partition_idx];
    }

    output[output_partition_idx].elements.resize(this_output_partition_size);
    if constexpr (keep_null_values) {
      output[output_partition_idx].null_values.resize(this_output_partition_size);
      null_values_as_char[output_partition_idx].resize(this_output_partition_size);
    }
  };

  constexpr auto OFFSET_JOB_SPAWN_THRESHOLD = 512;
  if (output_partition_count < OFFSET_JOB_SPAWN_THRESHOLD) {
    calculate_offsets();
  } else {
    jobs.emplace_back(std::make_shared<JobTask>(calculate_offsets));
  }
}

Hyrise::get().scheduler()->schedule_and_wait_for_tasks(jobs);
jobs.clear();
```

Greenscreen23 commented:
> Let's say we have 2^14 partitions. We have thousands of radix buckets (just obtaining the offset to write from the histogram aggregation is easily out of L1 cache). As we schedule per chunk, we end up with just a few 2/4/8 items per radix partition. No surprise we do not see any changes in the large radix bits.

Ah, that's annoying. But we cannot easily reduce the number of radix bits in these cases, because we want the hash table to fit in the cache, right?

We could look some stride ahead, calculate the hash of this element as well and prefetch the location. All integer types have a trivial hash, which is not expensive. This would be the code snippet we would have to insert at the start of the `for` loop going over the input chunk. We might even want to remove the software-managed buffers or only use them if the chunk size to radix bucket ratio is high enough.

```cpp
constexpr auto STRIDE = 16;
if (input_idx + STRIDE < elements_count) {
  const auto radix = hash_function(static_cast<HashedType>(elements[input_idx + STRIDE].value)) & radix_mask;
  __builtin_prefetch(output_offsets_by_input_partition[input_partition_idx].data() + radix, 0, 0);
  // or
  __builtin_prefetch(&tmp.data[radix].with_indices.count, 0, 0);
}
```

Bouncner commented:
Let’s talk on Monday. I think there are two issues.
First, our task per chunk model breaks as the OS “some form” of coordination. We need to group better.
Second, having vectors of vectors is just no a very good decision and easy to improve.

Greenscreen23 commented:
> Let’s talk on Monday. I think there are two issues. First, our task per chunk model breaks as the OS “some form” of coordination. We need to group better. Second, having vectors of vectors is just no a very good decision and easy to improve.

Grouping should be easy to implement, we would just have to tune it. To solve the vector of vectors problem, we could use an [mdspan](https://en.cppreference.com/w/cpp/container/mdspan.html). Or did you think of some other solution?

Bouncner commented:
Grouping is not per se hard, but I'd like to have a solution that works no only for the join. So maybe as part of the scheduler. Not sure.

I was thinking about `span`s, but not `mdspan`. That's new to me. But it looks exactly right. Awesome.

Bouncner commented:
But: it's C++23. :-/

Not sure how painful the switch is. Would be nice to up the C++ standard, but it would further block this PR.

Greenscreen23 commented:
> Grouping is not per se hard, but I'd like to have a solution that works no only for the join. So maybe as part of the scheduler. Not sure.

But we also have the issue that software-managed buffers are quite the overhead if every partition is only a few elements. I think grouping in this PR makes sense to try to minimize this overhead.

> I was thinking about `span`s, but not `mdspan`. That's new to me. But it looks exactly right. Awesome.

What would have been exactly right is an `mdvector`, but C++ is missing that. Doing all those things with spans is quite annoying, as the `materialize_input` function has to somehow return the underlying vector and the span array (or the mdspan if we have C++ 23).
